gtsfm/view_graph_estimator/reprojection_error_view_graph_estimator.py

```diff
@@ -0,0 +1,328 @@
+        wti_list, ta_metrics = self._trans_avg_module.run(num_images, i2Ui1_dict, wRi_list, gt_wTi_list=gt_wTi_list)
+
+        if gt_wTi_list:
+            ra_metrics = metrics_utils.compute_global_rotation_metrics(wRi_list, wti_list, gt_wTi_list)
+        else:
+            ra_metrics = None
+
+        init_cameras_dict = multi_view_optimizer.init_cameras(wRi_list, wti_list, calibrations)
+        ba_input_data, _ = self._data_association_module.run(
+            num_images=num_images,
+            cameras=init_cameras_dict,
+            corr_idxs_dict=corr_idxs_i1_i2,
+            keypoints_list=keypoints_list,
+            images=None,
+            cameras_gt=cameras_gt,
+        )
+        unfiltered_data, filtered_data = self._bundle_adjustment_module.run(ba_input_data)
+        reproj_errors = unfiltered_data.get_scene_reprojection_errors()
+        wTi_list = unfiltered_data.get_camera_poses()
+
+        tri_angle_percentiles = []
+        assert len(i2Ri1_dict.keys()) == 3
+        for (i1, i2) in i2Ri1_dict.keys():
+            if unfiltered_data.get_camera(i1) is None or unfiltered_data.get_camera(i2) is None:
+                tri_angle_percentiles.append(0)
+                continue
+            # TODO: try w/ filtered and unfiltered points.
+            # Ref: https://github.com/colmap/colmap/blob/dev/src/sfm/incremental_mapper.cc#L1065
+            tri_angles = mvs_utils.calculate_triangulation_angles_in_degrees(
+                camera_1=unfiltered_data.get_camera(i1),
+                camera_2=unfiltered_data.get_camera(i2),
+                points_3d=unfiltered_data.get_point_cloud(),
+            )
+            repr_percentile = np.percentile(tri_angles, 75)
+            tri_angle_percentiles.append(repr_percentile)
+            R_error_deg = two_view_reports[(i1, i2)].R_error_deg
+            U_error_deg = two_view_reports[(i1, i2)].U_error_deg
+            os.makedirs(os.path.join("plots", "tri_angle_histograms"), exist_ok=True)
+            plt.hist(tri_angles, bins=np.arange(100))
+            plt.title(f"75 Percentile: {repr_percentile:.1f}. R error: {R_error_deg:.1f} U error {U_error_deg:.1f}")
+            plt.savefig(os.path.join("plots", "tri_angle_histograms", f"{i1}_{i2}.jpg"), dpi=500)
+            plt.close("all")
+
+        # also check entropy of these histograms above
+        min_tri_angle = min(tri_angle_percentiles)
+
+        ba_metrics = self._bundle_adjustment_module.evaluate(unfiltered_data, filtered_data, cameras_gt)
+        return wTi_list, reproj_errors, min_tri_angle, ra_metrics, ta_metrics, ba_metrics
+
+    def optimize_three_views_incremental(self) -> None:
+        """
+        Estimate global camera poses for 3-views using PNP / absolute pose estimation.
+        """
+        import pycolmap
+
+        reconstructions = pycolmap.incremental_mapping(
+            database_path, image_dir, models_path, num_threads=min(multiprocessing.cpu_count(), 16)
+        )
```

Comment:
Since COLMAP is being called directly here, maybe add a reference to the underlying COLMAP function.

Reply:
Thanks for taking a look Travis. Sure, I'll add more details. This part is still WIP.